messaging/src/main/java/org/axonframework/lifecycle/ShutdownLatch.java

```diff
@@ -0,0 +1,98 @@
+package org.axonframework.lifecycle;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+
+/**
+ * A latch implementation to be used in shutdown scenarios. Operations to wait for can be added by invoking {@link
+ * #increment()}. An added operation should always be removed through {@link #decrement()} once it has completed, as
+ * {@link #await()} will otherwise block indefinitely. If the latch is waited on through {@link #await()}, new
+ * operations can no longer be added.
+ *
+ * @author Steven van Beelen
+ * @since 4.3
+ */
+public class ShutdownLatch {
+
+    private final AtomicInteger operationCounter = new AtomicInteger(0);
+    private CompletableFuture<Void> latch;
+
+    /**
+     * Add an operation this latch should wait on before opening up. If this operation is called whilst {@link #await()}
+     * is already called an {@link IllegalStateException} will be thrown.
+     *
+     * @throws IllegalStateException if {@link #await()} has been called prior to invoking this method
+     */
+    public void increment() {
+        if (isShuttingDown()) {
+            throw new IllegalStateException("No new operations can be added, since this latched is waited for.");
```

Comment:
This message won't mean much to people who are not familiar with the implementation details. Consider throwing a "ShutdownInProgressException" and/or change the message to "Cannot start activity. Shutdown in progress."